pkg/auth/authmap_gc.go

```diff
@@ -0,0 +1,203 @@
+		if err := r.cleanupMissingIdentities(); err != nil {
+			e.Done(err)
+			return fmt.Errorf("failed to cleanup missing identities: %w", err)
+		}
+	case resource.Delete:
+		log.Debugf("auth: identity deleted - cleaning up: %s", e.Key.Name)
+		if err := r.cleanupDeletedIdentity(e.Object); err != nil {
+			e.Done(err)
+			return fmt.Errorf("failed to cleanup deleted identity: %w", err)
+		}
+		r.discoveredCiliumIdentities = nil
+	}
+	e.Done(nil)
+	return nil
+}
+
+func (r *authMapGarbageCollector) cleanupMissingNodes() error {
+	return r.cleanupEntries(func(key authKey, info authInfo) bool {
+		if _, ok := r.discoveredCiliumNodeIDs[key.remoteNodeID]; !ok {
+			log.Debugf("auth: deleting entry due to removed remote node: %d", key.remoteNodeID)
+			return true
+		}
+		return false
+	})
+}
+
+func (r *authMapGarbageCollector) cleanupMissingIdentities() error {
+	return r.cleanupEntries(func(key authKey, info authInfo) bool {
+		if _, ok := r.discoveredCiliumIdentities[key.localIdentity]; !ok {
+			log.Debugf("auth: deleting entry due to removed local identity: %d", key.localIdentity)
+			return true
+		}
+		if _, ok := r.discoveredCiliumIdentities[key.remoteIdentity]; !ok {
+			log.Debugf("auth: deleting entry due to removed remote identity: %d", key.remoteIdentity)
+			return true
+		}
+		return false
+	})
+}
+
+func (r *authMapGarbageCollector) cleanupDeletedNode(node *ciliumv2.CiliumNode) error {
+	remoteNodeIDs := r.remoteNodeIDs(node)
+
+	return r.cleanupEntries(func(key authKey, info authInfo) bool {
+		for _, id := range remoteNodeIDs {
+			if key.remoteNodeID == id {
+				log.Debugf("auth: deleting entry due to removed node: %d", id)
+				return true
+			}
+		}
+		return false
+	})
+}
+
+func (r *authMapGarbageCollector) cleanupDeletedIdentity(id *ciliumv2.CiliumIdentity) error {
+	idNumeric, err := identity.ParseNumericIdentity(id.Name)
+	if err != nil {
+		return fmt.Errorf("failed to parse deleted identity: %w", err)
+	}
+
+	return r.cleanupEntries(func(key authKey, info authInfo) bool {
+		if key.localIdentity == idNumeric || key.remoteIdentity == idNumeric {
+			log.Debugf("auth: deleting entry due to removed identity: %d", idNumeric)
+			return true
+		}
+		return false
+	})
+}
+
+func (r *authMapGarbageCollector) CleanupExpiredEntries(_ context.Context) error {
+	log.Debug("auth: cleaning up expired entries")
+	now := time.Now()
+	err := r.cleanupEntries(func(key authKey, info authInfo) bool {
+		if info.expiration.Before(now) {
+			log.Debugf("auth: deleting entry due to expiration: %s", info.expiration)
+			return true
+		}
+		return false
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to cleanup expired entries: %w", err)
+	}
+	return nil
+}
+
+func (r *authMapGarbageCollector) cleanupEntries(predicate func(key authKey, info authInfo) bool) error {
+	all, err := r.authmap.All()
+	if err != nil {
+		return fmt.Errorf("failed to fetch all entries: %w", err)
+	}
+
+	var entriesToDelete []authKey
+	for k, v := range all {
```

Comment:
IMO this iteration needs to be done with the map locked?

Reply:
If locking the `all` map becomes too heavy, then maybe collect "gc event info" and iterate through the bpf map once per GC period, removing all the needed entries (expired, removed IDs, removed nodes) during one iteration?